memgpt/server/server.py

```diff
@@ -501,43 +504,85 @@ def user_message(self, user_id: uuid.UUID, agent_id: uuid.UUID, message: Union[s
+            # Convert to a Message object
+            message = Message(
+                user_id=user_id,
+                agent_id=agent_id,
+                role="user",
+                text=packaged_user_message,
+                # name=None,  # TODO handle name via API
+            )
+
+        if isinstance(message, Message):
+            # Can't have a null text field
+            if len(message.text) == 0 or message.text is None:
+                raise ValueError(f"Invalid input: '{message.text}'")
             # If the input begins with a command prefix, reject
             elif message.text.startswith("/"):
                 raise ValueError(f"Invalid input: '{message.text}'")
-            packaged_user_message = message
+
         else:
-            raise ValueError(f"Invalid input: '{message}'")
+            raise TypeError(f"Invalid input: '{message}' - type {type(message)}")
+
+        if timestamp:
+            # Override the timestamp with what the caller provided
+            message.created_at = timestamp
 
         # Run the agent state forward
         self._step(user_id=user_id, agent_id=agent_id, input_message=packaged_user_message)
 
     @LockingServer.agent_lock_decorator
-    def system_message(self, user_id: uuid.UUID, agent_id: uuid.UUID, message: str) -> None:
+    def system_message(
+        self, user_id: uuid.UUID, agent_id: uuid.UUID, message: Union[str, Message], timestamp: Optional[datetime] = None
+    ) -> None:
         """Process an incoming system message and feed it through the MemGPT agent"""
         if self.ms.get_user(user_id=user_id) is None:
             raise ValueError(f"User user_id={user_id} does not exist")
         if self.ms.get_agent(agent_id=agent_id, user_id=user_id) is None:
             raise ValueError(f"Agent agent_id={agent_id} does not exist")
 
         # Basic input sanitization
-        if not isinstance(message, str) or len(message) == 0:
-            raise ValueError(f"Invalid input: '{message}'")
+        if isinstance(message, str):
+            if len(message) == 0:
+                raise ValueError(f"Invalid input: '{message}'")
 
-        # If the input begins with a command prefix, reject
-        elif message.startswith("/"):
-            raise ValueError(f"Invalid input: '{message}'")
+            # If the input begins with a command prefix, reject
+            elif message.startswith("/"):
+                raise ValueError(f"Invalid input: '{message}'")
 
-        # Else, process it as a user message to be fed to the agent
-        else:
-            # Package the user message first
             packaged_system_message = system.package_system_message(system_message=message)
-            # Run the agent state forward
-            self._step(user_id=user_id, agent_id=agent_id, input_message=packaged_system_message)
+
+            # NOTE: eventually deprecate and only allow passing Message types
+            # Convert to a Message object
+            message = Message(
```

Comment:
@sarahwooders now when we call `Agent.step()` from `server.py`, we should only be sending `Message` types in, no more `str` types